File: android/src/main/java/expo/modules/mediapipefacedetector/ExpoMediapipeFaceDetectorView.kt
package expo.modules.mediapipefacedetector

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.util.Size
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.framework.image.MPImage
import com.google.mediapipe.tasks.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.facedetector.FaceDetector
import com.google.mediapipe.tasks.vision.facedetector.FaceDetectorOptions
import com.google.mediapipe.tasks.vision.facedetector.FaceDetectorResult
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class ExpoMediapipeFaceDetectorView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  private val onFaceDetected by EventDispatcher()
  private val onError by EventDispatcher()

  private val previewView = PreviewView(context).apply {
    layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
  }

  private var cameraProvider: ProcessCameraProvider? = null
  private var preview: Preview? = null
  private var imageAnalyzer: ImageAnalysis? = null
  private var camera: Camera? = null
  private var faceDetector: FaceDetector? = null
  
  private lateinit var cameraExecutor: ExecutorService
  private var isEnabled = true
  private var currentConfig: Map<String, Any>? = null

  init {
    addView(previewView)
    cameraExecutor = Executors.newSingleThreadExecutor()
    setupFaceDetector()
  }

  fun updateConfig(config: Map<String, Any>?) {
    currentConfig = config
    setupFaceDetector()
  }

  fun setEnabled(enabled: Boolean) {
    isEnabled = enabled
    if (enabled) {
      startCamera()
    } else {
      stopCamera()
    }
  }

  private fun setupFaceDetector() {
    try {
      val optionsBuilder = FaceDetectorOptions.builder()

      currentConfig?.let { config ->
        val minDetectionConfidence = config["minDetectionConfidence"] as? Double ?: 0.5
        val minSuppressionThreshold = config["minTrackingConfidence"] as? Double ?: 0.5
        
        optionsBuilder.setMinDetectionConfidence(minDetectionConfidence.toFloat())
        optionsBuilder.setMinSuppressionThreshold(minSuppressionThreshold.toFloat())
        optionsBuilder.setRunningMode(RunningMode.LIVE_STREAM)
        
        // Set result listener for live stream mode
        optionsBuilder.setResultListener { result, _ ->
          handleDetectionResult(result)
        }
        
        optionsBuilder.setErrorListener { error, _ ->
          onError(mapOf(
            "message" to "Face detection error: ${error.message}",
            "code" to "DETECTION_ERROR"
          ))
        }
      }

      // Set base options with model
      val baseOptionsBuilder = BaseOptions.builder()
      baseOptionsBuilder.setModelAssetPath("face_detection_short_range.tflite")
      optionsBuilder.setBaseOptions(baseOptionsBuilder.build())

      faceDetector = FaceDetector.createFromOptions(context, optionsBuilder.build())

    } catch (e: Exception) {
      onError(mapOf(
        "message" to "Failed to initialize face detector: ${e.message}",
        "code" to "INITIALIZATION_ERROR"
      ))
    }
  }

  private fun startCamera() {
    if (!isEnabled) return

    val cameraProviderFuture = ProcessCameraProvider.getInstance(context)
    cameraProviderFuture.addListener({
      try {
        cameraProvider = cameraProviderFuture.get()
        bindCameraUseCases()
      } catch (e: Exception) {
        onError(mapOf(
          "message" to "Failed to start camera: ${e.message}",
          "code" to "CAMERA_ERROR"
        ))
      }
    }, ContextCompat.getMainExecutor(context))
  }

  private fun bindCameraUseCases() {
    val cameraProvider = cameraProvider ?: return
    val lifecycleOwner = context as? LifecycleOwner ?: return

    // Preview use case
    preview = Preview.Builder()
      .setTargetResolution(Size(640, 480))
      .build()
      .also {
        it.setSurfaceProvider(previewView.surfaceProvider)
      }

    // Image analysis use case for face detection
    imageAnalyzer = ImageAnalysis.Builder()
      .setTargetResolution(Size(640, 480))
      .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
      .build()
      .also {
        it.setAnalyzer(cameraExecutor) { imageProxy ->
          detectFaces(imageProxy)
        }
      }

    // Select front camera
    val cameraSelector = CameraSelector.DEFAULT_FRONT_CAMERA

    try {
      // Unbind use cases before rebinding
      cameraProvider.unbindAll()

      // Bind use cases to camera
      camera = cameraProvider.bindToLifecycle(
        lifecycleOwner,
        cameraSelector,
        preview,
        imageAnalyzer
      )

    } catch (e: Exception) {
      onError(mapOf(
        "message" to "Failed to bind camera use cases: ${e.message}",
        "code" to "CAMERA_BIND_ERROR"
      ))
    }
  }

  private fun detectFaces(imageProxy: ImageProxy) {
    val detector = faceDetector ?: return

    try {
      val bitmap = imageProxy.toBitmap()
      val mpImage = BitmapImageBuilder(bitmap).build()
      val timestampMs = System.currentTimeMillis()

      detector.detectAsync(mpImage, timestampMs)
    } catch (e: Exception) {
      onError(mapOf(
        "message" to "Face detection error: ${e.message}",
        "code" to "DETECTION_ERROR"
      ))
    } finally {
      imageProxy.close()
    }
  }

  private fun handleDetectionResult(result: FaceDetectorResult) {
    val faces = result.detections().map { detection ->
      val boundingBox = detection.boundingBox()
      
      mapOf(
        "boundingBox" to mapOf(
          "x" to boundingBox.left,
          "y" to boundingBox.top,
          "width" to boundingBox.width(),
          "height" to boundingBox.height()
        ),
        "confidence" to (detection.categories().firstOrNull()?.score() ?: 0.0f),
        "landmarks" to (detection.keypoints()?.map { keypoint ->
          mapOf(
            "type" to "UNKNOWN",
            "position" to mapOf(
              "x" to keypoint.x(),
              "y" to keypoint.y(),
              "z" to (keypoint.z() ?: 0.0f)
            )
          )
        } ?: emptyList<Map<String, Any>>())
      )
    }

    onFaceDetected(mapOf(
      "faces" to faces,
      "imageWidth" to previewView.width,
      "imageHeight" to previewView.height,
      "timestamp" to System.currentTimeMillis()
    ))
  }

  private fun stopCamera() {
    cameraProvider?.unbindAll()
  }

  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    cameraExecutor.shutdown()
  }
}

File: ios/ExpoMediapipeFaceDetectorView.swift
import ExpoModulesCore
import UIKit
import AVFoundation
import MediaPipeTasksVision

// This view will be used as a native component for face detection preview
class ExpoMediapipeFaceDetectorView: ExpoView {
  private let previewLayer = AVCaptureVideoPreviewLayer()
  private let onFaceDetected = EventDispatcher()
  private let onError = EventDispatcher()
  
  private var captureSession: AVCaptureSession?
  private var videoOutput: AVCaptureVideoDataOutput?
  private var faceDetector: FaceDetector?
  private var isEnabled = true
  private var currentConfig: [String: Any]?
  
  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    setupPreviewLayer()
  }
  
  override func layoutSubviews() {
    super.layoutSubviews()
    previewLayer.frame = bounds
  }
  
  // MARK: - Public Methods
  
  func updateConfig(config: [String: Any]?) {
    currentConfig = config
    setupFaceDetector()
  }
  
  func setEnabled(enabled: Bool) {
    isEnabled = enabled
    if enabled {
      startCamera()
    } else {
      stopCamera()
    }
  }
  
  // MARK: - Private Methods
  
  private func setupPreviewLayer() {
    previewLayer.videoGravity = .resizeAspectFill
    layer.addSublayer(previewLayer)
  }
  
  private func setupFaceDetector() {
    do {
      let options = FaceDetectorOptions()
      
      if let config = currentConfig {
        options.minDetectionConfidence = config["minDetectionConfidence"] as? Float ?? 0.5
        options.minSuppressionThreshold = config["minTrackingConfidence"] as? Float ?? 0.5
        options.runningMode = .liveStream
        
        // Set up result callback for live stream mode
        options.faceDetectorLiveStreamDelegate = self
      }
      
      // Set model path
      if let modelPath = Bundle.main.path(forResource: "face_detection_short_range", ofType: "tflite") {
        options.baseOptions.modelAssetPath = modelPath
      }
      
      faceDetector = try FaceDetector(options: options)
    } catch {
      onError([
        "message": "Failed to initialize face detector: \(error.localizedDescription)",
        "code": "INITIALIZATION_ERROR"
      ])
    }
  }
  
  private func startCamera() {
    guard isEnabled else { return }
    
    captureSession = AVCaptureSession()
    guard let captureSession = captureSession else { return }
    
    captureSession.beginConfiguration()
    captureSession.sessionPreset = .high
    
    // Add camera input
    guard let camera = AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .front),
          let input = try? AVCaptureDeviceInput(device: camera) else {
      onError([
        "message": "Failed to access camera",
        "code": "CAMERA_ACCESS_ERROR"
      ])
      return
    }
    
    if captureSession.canAddInput(input) {
      captureSession.addInput(input)
    }
    
    // Add video output
    videoOutput = AVCaptureVideoDataOutput()
    guard let videoOutput = videoOutput else { return }
    
    videoOutput.setSampleBufferDelegate(self, queue: DispatchQueue(label: "camera.frame.processing.queue"))
    videoOutput.alwaysDiscardsLateVideoFrames = true
    videoOutput.videoSettings = [
      kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA
    ]
    
    if captureSession.canAddOutput(videoOutput) {
      captureSession.addOutput(videoOutput)
    }
    
    previewLayer.session = captureSession
    captureSession.commitConfiguration()
    
    DispatchQueue.global(qos: .userInitiated).async {
      captureSession.startRunning()
    }
  }
  
  private func stopCamera() {
    captureSession?.stopRunning()
    captureSession = nil
    videoOutput = nil
  }
}

// MARK: - AVCaptureVideoDataOutputSampleBufferDelegate

extension ExpoMediapipeFaceDetectorView: AVCaptureVideoDataOutputSampleBufferDelegate {
  func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
    guard isEnabled, let faceDetector = faceDetector else { return }
    
    do {
      let mpImage = try MPImage(sampleBuffer: sampleBuffer)
      let timestamp = CMSampleBufferGetPresentationTimeStamp(sampleBuffer)
      let timestampMs = Int(CMTimeGetSeconds(timestamp) * 1000)
      
      // Detect faces asynchronously
      try faceDetector.detectAsync(image: mpImage, timestampInMilliseconds: timestampMs)
    } catch {
      DispatchQueue.main.async {
        self.onError([
          "message": "Face detection error: \(error.localizedDescription)",
          "code": "DETECTION_ERROR"
        ])
      }
    }
  }
}

// MARK: - FaceDetectorLiveStreamDelegate

extension ExpoMediapipeFaceDetectorView: FaceDetectorLiveStreamDelegate {
  func faceDetector(_ faceDetector: FaceDetector, didFinishDetection result: FaceDetectorResult?, timestampInMilliseconds: Int, error: Error?) {
    if let error = error {
      DispatchQueue.main.async {
        self.onError([
          "message": "Face detection error: \(error.localizedDescription)",
          "code": "DETECTION_ERROR"
        ])
      }
      return
    }
    
    guard let result = result else { return }
    
    DispatchQueue.main.async {
      let formattedResult = self.formatDetectionResult(result: result, timestamp: timestampInMilliseconds)
      self.onFaceDetected(formattedResult)
    }
  }
  
  private func formatDetectionResult(result: FaceDetectorResult, timestamp: Int) -> [String: Any] {
    var faces: [[String: Any]] = []
    
    for detection in result.detections {
      var face: [String: Any] = [:]
      
      // Bounding box
      let boundingBox = detection.boundingBox
      face["boundingBox"] = [
        "x": boundingBox.origin.x,
        "y": boundingBox.origin.y,
        "width": boundingBox.size.width,
        "height": boundingBox.size.height
      ]
      
      // Confidence
      if let category = detection.categories.first {
        face["confidence"] = category.score
      }
      
      // Landmarks (if available)
      var landmarks: [[String: Any]] = []
      if let keypoints = detection.keypoints {
        for keypoint in keypoints {
          landmarks.append([
            "type": "UNKNOWN", // You'd implement proper landmark type mapping here
            "position": [
              "x": keypoint.location.x,
              "y": keypoint.location.y,
              "z": keypoint.location.z ?? 0
            ]
          ])
        }
      }
      face["landmarks"] = landmarks
      
      faces.append(face)
    }
    
    return [
      "faces": faces,
      "imageWidth": bounds.width,
      "imageHeight": bounds.height,
      "timestamp": timestamp
    ]
  }
}
